Metadata API: test invalid input values for attributes

MVrachev opened the pull request:
Fixes #1434

**Description of the changes being introduced by the pull request**:

A while ago we decided that it's best to research each of the individuals
attributes one by one and identify what level of validation it needs
compared to how we use it:
#1366 (comment).

This work is ongoing and there are a couple of commits already merged
for this:
- 6c5d970
- f20664d
- 41afb1e

We want to be able to test the validation we add for attributes against known bad
values. The way we want to do that is with table testing we have added
using decorators for our metadata classes defined in New API:
#1416. 
This gives us an easy way to add new cases for each of the attributes and
not depend on external files.

_**It's important to note that I haven't added invalid tests for all attributes, because
we haven't finished with the metadata research.**_

When new validation for a certain attribute is added we should make sure to add
invalid tests in `TestInvalidSerialization` as well.

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
Test new API: serialization tests for invalid arg

A while ago we decided that it's best to research each of the individuals
attributes one by one and identify what level of validation it needs
compared to how we use it:
https://github.com/theupdateframework/tuf/pull/1366#issuecomment-829288790.

This work is ongoing and there are a couple of commits already merged
for this:
- https://github.com/theupdateframework/tuf/commit/6c5d970799ffe1f79d02e24a8ea56afeae4e636b
- https://github.com/theupdateframework/tuf/commit/f20664d2fc967852d8e3aca7f951e4d1e7855e61
- https://github.com/theupdateframework/tuf/commit/41afb1e1342d5f8870180ba1ecea8894c03aec24

We want to be able to test the attributes validation against known bad
values.
The way we want to do that is with table testing we have added
using decorators for our metadata classes defined in New API:
https://github.com/theupdateframework/tuf/pull/1416.
This gives us an easy way to add new cases for each of the attributes and
not depend on external files.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
This pr is ready for review after the merge of #1416.

jku reviewed (changes requested):
This seems very fine grained. There's no benefit from having a separate test for each attribute (when tests are practically identical) but it makes the patch massive. I would expect less granularity would mean at least 50% reduction in commit size?

Also, currently it's hard to see what is being tested: I need to read deep into the test code to find what is the actual class that is being tested by an individual test.

I don't see why we would not do the exact same thing we already do for valid data tests: one invalid-data dataset and one new test per class-under-test. I also think the invalid test+data for a class should be next to the valid test+data for the same class -- If the tests need to be split by for some reason, I expect it to be more useful to split by class-under-test (so e.g. TestKeySerialization) than splitting by TestSerialization/TestInvalidSerialization: I often want to run tests for a specific class, but I have never wanted to run tests based whether success or failure is expected.

MVrachev commented:
> This seems very fine grained. There's no benefit from having a separate test for each attribute (when tests are practically identical) but it makes the patch massive. I would expect less granularity would mean at least 50% reduction in commit size?
> 
> Also, currently it's hard to see what is being tested: I need to read deep into the test code to find what is the actual class that is being tested by an individual test.
> 
> I don't see why we would not do the exact same thing we already do for valid data tests: one invalid-data dataset and one new test per class-under-test. I also think the invalid test+data for a class should be next to the valid test+data for the same class -- If the tests need to be split by for some reason, I expect it to be more useful to split by class-under-test (so e.g. TestKeySerialization) than splitting by TestSerialization/TestInvalidSerialization: I often want to run tests for a specific class, but I have never wanted to run tests based whether success or failure is expected.

I did it more granularly because we have more cases per attribute.
Will create a commit making test cases per class and we will comment it again.

MVrachev commented:
In a discussion with Jussi, he proposed that I should group invalid and valid tests close to each other in the same test class.
Added a commit doing that.

jku commented on tests/test_metadata_serialization.py:
```diff
@@ -62,6 +117,41 @@ def test_key_serialization(self, test_case_data: str):
+        "no keytype": '{"keyid": "id", "scheme": "rsassa-pss-sha256", "keyval": {"public": "foo"}}',
+        "no scheme": '{"keyid": "id", "keytype": "rsa", "keyval": {"public": "foo"}}',
+        "no keyval": '{"keyid": "id", "keytype": "rsa", "scheme": "rsassa-pss-sha256"}',
+        "no public in keyval": '{"keyid": "id", "keytype": "rsa", "scheme": "rsassa-pss-sha256", "keyval": {}}',
```
I believe this should not be invalid (I mean rsa keys do require a "public" but key in general should not be required to have one, this is my reading of the spec)

Maybe file an issue on this?

MVrachev replied:
I added it because this is the current behavior of the code.
Will remove it and create a pr about it.

MVrachev replied:
https://github.com/theupdateframework/tuf/pull/1480

jku commented on tests/test_metadata_serialization.py:
```diff
@@ -92,6 +182,25 @@ def test_root_serialization(self, test_case_data: str):
+        "length 0": '{"version": 1, "length": 0, "hashes": {"sha256" : "abc"}}',
+        "length below 0": '{"version": 1, "length": -1, "hashes": {"sha256" : "abc"}}',
+        "empty hashes dict": '{"version": 1, "length": 1, "hashes": {}}',
+        # Hashes keys cannot be strings because the JSON specification.
+        # Then, do we want to test for that?
```
this does not seem to belong in the test data (but I also don't understand the comment)

MVrachev replied:
Oh... Sorry, it was a comment to myself and forgot to remove it.

jku commented on tests/test_metadata_serialization.py:
```diff
@@ -62,6 +117,41 @@ def test_key_serialization(self, test_case_data: str):
+        "keytype wrong type": '{"keyid": "id", "keytype": 1, "scheme": "rsassa-pss-sha256", "keyval": {"public": "abc"}}',
+        "scheme wrong type": '{"keyid": "id", "keytype": "rsa", "scheme": 1, "keyval": {"public": "abc"}}',
+        "keyval wrong type": '{"keyid": "id", "keytype": "rsa", "scheme": "rsassa-pss-sha256", "keyval": 1}',
+        "keyval public wrong type": '{"keyid": "id", "keytype": "rsa", "scheme": "rsassa-pss-sha256", "keyval": {"public": 1}}',
```
same as above: this should not be invalid for key in general

MVrachev replied:
Removed it and filed a pr: https://github.com/theupdateframework/tuf/pull/1480

jku commented on tests/test_metadata_serialization.py:
```diff
@@ -48,8 +50,61 @@ def wrapper(test_cls: "TestSerialization"):
+
 class TestSerialization(unittest.TestCase):
 
+    invalid_signed: DataSet = {
```
there's a lot of test cases here that are unlikely to be useful (e.g. 5 invalid spec versions  is just overkill) but since the tests are so fast... I guess no harm done?

MVrachev replied:
I removed one of the cases one digit spec_version, the others seem relevant to the checks we are making.

jku replied:
ok, but do keep in mind that the idea isn't to test the lines of code that we happen to have in there right now: the idea is to have test data that covers as much of the possible inputs as practically useful -- since we can't cover 100% of the possibilities we inevitably end up making decisions on what's important: we test the inputs we think could be interesting.

jku commented on tests/test_metadata_serialization.py:
```diff
@@ -48,8 +50,61 @@ def wrapper(test_cls: "TestSerialization"):
     return real_decorator
 
 
+class TestSigned(Signed):
```
I think creating a class just for testing it feels wrong. I would prefer using an existing Signed-derivative for testing these fields, but it's your choice.

MVrachev replied:
I removed one of the cases `one digit spec_version`, the others seem relevant to the checks we are making.

MVrachev replied:
I will remove it, but as a side effect, the `invalid_signed` tests will become longer.

jku reviewed (commented):
I would still really like it if the variable and function names were consistent

> invalid_key vs valid_key**s**
> invalid_role vs valid_role**s**
> invalid_metafile_attr vs valid_metafiles
> invalid_targetfile_hashes_length vs valid_targetfiles

With some of those I really need to squint to figure out that they are supposed to be tests for the same class

Other comments are just nitpicks I think

MVrachev commented:
@jku squashed my commits and addressed all of your comments (including the inconsistencies) except for the ones
about `keyval` which are addressed in another pr.

jku commented on tests/test_metadata_serialization.py:
```diff
@@ -159,6 +247,20 @@ def test_delegation_serialization(self, test_case_data: str):
 
+    invalid_targetfiles: DataSet = {
+        "no hashes": '{"length": 1}',
+        "no length": '{"hashes": {"sha256": 1}}'
```
that's an invalid hashes value type

jku commented on tests/test_metadata_serialization.py:
```diff
@@ -50,18 +52,87 @@ def wrapper(test_cls: "TestSerialization"):
+        "no expires": '{"_type": "signed", "spec_version": "1.0.0", "version": 1, "meta": {"f.txt": {"version": 1}}}',
+        "empty str _type": '{"_type": "", "spec_version": "1.0.0", "version": 1, "expires": "2030-01-01T00:00:00Z", \
+            "meta": {"f.txt": {"version": 1}}}',
+        "empty str spec_version": '{"_type": "signed", "spec_version": "", "version": 1, "expires": "2030-01-01T00:00:00Z", \
+            "meta": {"f.txt": {"version": 1}}}',
```
Snapshot meta isn't required to contain anything: `"meta": {}` is enough if you're going for the minimal serialised form

nitpick on the style: I think this would be easier to read if you just barely need two lines (also allows names to be as long as you want):
```
        "empty str spec_version":
            '{"_type": "signed", "spec_version": "", "version": 1, "expires": "2030-01-01T00:00:00Z", "meta": {}}',
```

MVrachev replied:
Okay, will do that, but will add a comment that per spec having a snapshot object like this in a repository is considered invalid.

jku reviewed (approved):
LGTM, left a couple of comments that are up to you

jku commented:
(that last comment is a reply to earlier discussion on spec_version test cases, for some reason github shows it independently here)

MVrachev commented:
@jku I addressed your comments.

Commit:
Metadata API: fix Key validation exception

In pr https://github.com/theupdateframework/tuf/pull/1449 we introduced
Key validation and there decided to raise "ValueError" if one of
keyid, scheme, keyval or keytype is not a string.
That seems like a mistake given that in the python docs it's written:
"Passing arguments of the wrong type
(e.g. passing a list when an int is expected) should result
in a TypeError, but passing arguments with the wrong value
(e.g. a number outside expected boundaries) should result
in a ValueError."

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
Rebased on top of the latest changes in develop branch and fixed conflicts.
